File: OC/XWMP3ToWAV/XWMP3ToWAV/XWMP3ToPCM.m
//
//  XWMP3ToPCM.m
//  XWMP3ToWAV
//
//  Created by 邱学伟 on 2017/3/13.
//  Copyright © 2017年 邱学伟. All rights reserved.
//

#import "XWMP3ToPCM.h"

#import <AVFoundation/AVFoundation.h>

@implementation XWMP3ToPCM
/// 用mp3文件创建一个AVAsset实例
+ (AVAsset *)readMp3FileWithMp3FilePath:(NSString *)mp3FilePath {
    AVAsset *asset = [AVAsset assetWithURL:[NSURL fileURLWithPath:mp3FilePath]];
    if (!asset) {
        NSLog(@"不存在MP3文件!");
        return NULL;
    }
    return asset;
}

//// 用一个AVAsset实例创建一个AVAssetReader实例
+ (AVAssetReader *)initAssetReader:(AVAsset *)avasset {
    
    NSError *error = nil;
    AVAssetReader *assetReader = [AVAssetReader assetReaderWithAsset:avasset error:&error];
    if (error) {
        NSLog(@"initAssetReader - error:%@",error);
    }
    return assetReader;
}

/// 创建一个AVAssetWriter实例
+ (AVAssetWriter *)initAssetWriterWithPcmFilePath:(NSString *)pcmFilePath {
    
    NSError *error = nil;
    AVAssetWriter *assetWriter = [AVAssetWriter assetWriterWithURL:[NSURL fileURLWithPath:pcmFilePath] fileType:AVFileTypeWAVE error:&error];
    if (error) {
        NSLog(@"initAssetWriterWithPcmFilePath -- error:%@",error);
    }
    return assetWriter;
}

+ (void)mp3ToPcmWithMp3FilePath:(NSString *)mp3FilePath pcmFilePath:(NSString *)pcmFilePath {
    
    NSLog(@"+++ mp3FilePath:\(%@) -> pcmFilePath:\(%@)",mp3FilePath,pcmFilePath);
    
    // 1.创建AVAsset实例
    AVAsset *asset = [self readMp3FileWithMp3FilePath:mp3FilePath];
    // 2.创建AVAssetReader实例
    AVAssetReader *assetReader = [self initAssetReader:asset];
    
   // 3.配置转码参数
    AudioChannelLayout channelLayout;// =  AudioChannelLayout()
    memset(&channelLayout, 0, sizeof(AudioChannelLayout));
    channelLayout.mChannelLayoutTag = kAudioChannelLayoutTag_Stereo;
    
    [NSData dataWithBytes:&channelLayout length:sizeof(AudioChannelLayout)];

    /*
         AVFormatIDKey :  @"lpcm",    // 音频格式
         AVSampleRateKey : [NSNumber numberWithFloat:44100.0],    // 采样率
         AVNumberOfChannelsKey : @2,    // 通道数 1 || 2
         AVChannelLayoutKey : [NSData dataWithBytes:&channelLayout length:sizeof(AudioChannelLayout)],  // 声音效果（立体声）
         AVLinearPCMBitDepthKey : @16  // 音频的每个样点的位数
         AVLinearPCMIsNonInterleaved : noNumber,  // 音频采样是否非交错
         AVLinearPCMIsFloatKey : noNumber,    // 采样信号是否浮点数
         AVLinearPCMIsBigEndianKey : noNumber // 音频采用高位优先的记录格式
     */
    NSNumber *noNumber = [NSNumber numberWithBool:NO];
    NSMutableDictionary *outputSettings = [[NSMutableDictionary alloc] init];
    [outputSettings setValue:[NSNumber numberWithInt:kAudioFormatLinearPCM] forKey:AVFormatIDKey];
    [outputSettings setValue:[NSNumber numberWithFloat:44100.0] forKey:AVSampleRateKey];
    [outputSettings setValue:[NSNumber numberWithInt:2] forKey:AVNumberOfChannelsKey];
    [outputSettings setValue:[NSData dataWithBytes:&channelLayout length:sizeof(AudioChannelLayout)] forKey:AVChannelLayoutKey];
    [outputSettings setValue:[NSNumber numberWithInt:16] forKey:AVLinearPCMBitDepthKey];
    [outputSettings setValue:noNumber forKey:AVLinearPCMIsNonInterleaved];
    [outputSettings setValue:noNumber forKey:AVLinearPCMIsFloatKey];
    [outputSettings setValue:noNumber forKey:AVLinearPCMIsBigEndianKey];
    
    AVAssetReaderAudioMixOutput *readerAudioMixOutput = [AVAssetReaderAudioMixOutput assetReaderAudioMixOutputWithAudioTracks:asset.tracks audioSettings:outputSettings];
    
    // 4.创建AVAssetReaderAudioMixOutput实例并绑定到assetReader上
    if (![assetReader canAddOutput:readerAudioMixOutput]) {
        
        NSLog(@"can't add readerAudioMixOutput");
        return;
    }
    [assetReader addOutput:readerAudioMixOutput];
    
    // MARK: 1.保存到文件
    // 5.创建AVAssetWriter实例
    AVAssetWriter *assetWriter = [self initAssetWriterWithPcmFilePath:pcmFilePath];
    
    // 6.创建AVAssetWriterInput实例并绑定到assetWriter上
    if (![assetWriter canApplyOutputSettings:outputSettings forMediaType:AVMediaTypeAudio]) {
        
        NSLog(@"can't apply outputSettings");
        return;
    }
    
    AVAssetWriterInput *writerInput = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeAudio outputSettings:outputSettings];
    // 是否让媒体数据保持实时。在此不需要开启
    writerInput.expectsMediaDataInRealTime = NO;
    
    
    if (![assetWriter canAddInput:writerInput]) {
        NSLog(@"can't add writerInput");
        return;
    }
    [assetWriter addInput:writerInput];
    
    // 7.启动转码
    [assetReader startReading];
    [assetWriter startWriting];
    
    // 开启session
    AVAssetTrack *track = asset.tracks.firstObject;
    if (!track) { return; }
    
    CMTime startTime = CMTimeMakeWithSeconds(0, track.naturalTimeScale);
    
    [assetWriter startSessionAtSourceTime:startTime];
    
    [writerInput requestMediaDataWhenReadyOnQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0) usingBlock:^{
        while ([writerInput isReadyForMoreMediaData]) {
            CMSampleBufferRef nextBuffer = [readerAudioMixOutput copyNextSampleBuffer];
            if (nextBuffer) {
                [writerInput appendSampleBuffer:nextBuffer];
            }else{
                [writerInput markAsFinished];
                [assetReader cancelReading];
                [assetWriter finishWritingWithCompletionHandler:^{
                    NSLog(@"whrite complete");
                }];
                break;
            }
        }
    }];
}


@end

File: Swift/XWMP3ToWAV/XWMP3ToWAV/XWMP3ToPCM.swift
//
//  XWMP3ToPCM.swift
//  XWMP3ToPCM
//
//  Created by 邱学伟 on 2017/3/13.
//  Copyright © 2017年 邱学伟. All rights reserved.
//

import UIKit
import AVFoundation

class XWMP3ToPCM: NSObject { }

//MARK: - public
extension XWMP3ToPCM {
    
    /// 传入 MP3 文件路径 和 需要存储pcm文件的文件路径 注意 pcmFilePath 后缀为 .wav
    class func mp3ToPcm(mp3FilePath : String, pcmFilePath : String)  {
        
        print("+++ mp3FilePath:\(mp3FilePath) -> pcmFilePath:\(pcmFilePath)")
        
        // MARK: 通用逻辑
        // 1.创建AVAsset实例
        guard let asset = self.readMp3File(mp3FilePath) else { return }
        
        // 2.创建AVAssetReader实例
        guard let assetReader = self.initAssetReader(asset: asset) else { return }
        
        // 3.配置转码参数
        var channelLayout = AudioChannelLayout()
        memset(&channelLayout, 0, MemoryLayout<AudioChannelLayout>.size)
        channelLayout.mChannelLayoutTag = kAudioChannelLayoutTag_Stereo;
        
        let outputSettings = [
            AVFormatIDKey : kAudioFormatLinearPCM,    // 音频格式
            AVSampleRateKey : 44100.0,    // 采样率
            AVNumberOfChannelsKey : 2,    // 通道数 1 || 2
            AVChannelLayoutKey : Data.init(bytes: &channelLayout, count: MemoryLayout<AudioChannelLayout>.size),  // 声音效果（立体声）
            AVLinearPCMBitDepthKey : 16,  // 音频的每个样点的位数
            AVLinearPCMIsNonInterleaved : false,  // 音频采样是否非交错
            AVLinearPCMIsFloatKey : false,    // 采样信号是否浮点数
            AVLinearPCMIsBigEndianKey : false // 音频采用高位优先的记录格式
            ] as [String : Any]
        
        
        // 4.创建AVAssetReaderAudioMixOutput实例并绑定到assetReader上
        let readerAudioMixOutput = AVAssetReaderAudioMixOutput(audioTracks: asset.tracks, audioSettings: outputSettings)
        
        if !assetReader.canAdd(readerAudioMixOutput) {
            
            print("can't add readerAudioMixOutput")
            return
        }
        
        assetReader.add(readerAudioMixOutput)
        
        // MARK: 1.保存到文件
        
        // 5.创建AVAssetWriter实例
        guard let assetWriter = self.initAssetWriter(pcmFilePath) else { return }
        
        // 6.创建AVAssetWriterInput实例并绑定到assetWriter上
        
        if !assetWriter.canApply(outputSettings: outputSettings, forMediaType: AVMediaTypeAudio) {
            
            print("can't apply outputSettings")
            return
        }
        
        let writerInput = AVAssetWriterInput(mediaType: AVMediaTypeAudio, outputSettings: outputSettings)
        
        // 是否让媒体数据保持实时。在此不需要开启
        writerInput.expectsMediaDataInRealTime = false
        
        if !assetWriter.canAdd(writerInput) {
            
            print("can't add writerInput")
            return
        }
        
        assetWriter.add(writerInput)
        
        // 7.启动转码
        assetReader.startReading()
        assetWriter.startWriting()
        
        // 开启session
        guard let track = asset.tracks.first else { return }
        let startTime = CMTime(seconds: 0, preferredTimescale: track.naturalTimeScale)
        assetWriter.startSession(atSourceTime: startTime)
        
        let mediaInputQueue = DispatchQueue(label: "mediaInputQueue")
        writerInput.requestMediaDataWhenReady(on: mediaInputQueue, using: {
            
            while writerInput.isReadyForMoreMediaData {
                
                if let nextBuffer = readerAudioMixOutput.copyNextSampleBuffer() {
                    writerInput.append(nextBuffer)
                    
                } else {
                    
                    writerInput.markAsFinished()
                    assetReader.cancelReading()
                    assetWriter.finishWriting(completionHandler: {
                        print("whrite complete")
                    })
                    
                    break
                }
            }
        })
        
        // MARK: 2.转成NSDate保存在内存中
        
        //        // 5.启动转码
        //        assetReader.startReading()
        //        var PCMData = Data()
        //
        //        while let nextBuffer = readerAudioMixOutput.copyNextSampleBuffer() {
        //
        //            var audioBufferList = AudioBufferList()
        //            var blockBuffer: CMBlockBuffer?
        //
        //            // CMSampleBuffer to Data
        //            CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(nextBuffer,
        //                                                                    nil,
        //                                                                    &audioBufferList,
        //                                                                    MemoryLayout<AudioBufferList>.size,
        //                                                                    nil,
        //                                                                    nil,
        //                                                                    0,
        //                                                                    &blockBuffer)
        //
        //            let audioBuffer = audioBufferList.mBuffers
        //            guard let frame = audioBuffer.mData else { continue }
        //
        //            PCMData.append(frame.assumingMemoryBound(to: UInt8.self), count: Int(audioBuffer.mDataByteSize))
        //            blockBuffer = nil;
        //        }
        //        
        //        print("whrite complete")
    }
}

//MARK: - private
extension XWMP3ToPCM {
    
    // 用mp3文件创建一个AVAsset实例
    class func readMp3File(_ mp3FilePath : String)  -> AVAsset? {
        
//        guard let mp3FilePath = Bundle.main.path(forResource: "trust you", ofType: "mp3") else { return nil }
        let fileURL = URL(fileURLWithPath: mp3FilePath)
        let asset = AVAsset(url: fileURL)
        print("asset.tracks : %zd", asset.tracks.count)
        return asset
    }
    
    // 用一个AVAsset实例创建一个AVAssetReader实例
    class func initAssetReader(asset: AVAsset) -> AVAssetReader? {
        
        let assetReader: AVAssetReader
        
        do {
            assetReader = try AVAssetReader(asset: asset)
            
        } catch {
            
            print(error)
            return nil
        }
        
        return assetReader
    }
    
    // 创建一个AVAssetWriter实例
    class func initAssetWriter(_ pcmFilePath : String) -> AVAssetWriter? {
        
        let assetWriter: AVAssetWriter
        
//        guard let outPutPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first else { return nil }
//        // 这里的扩展名'.wav'只是标记了文件的打开方式，实际的编码封装格式由assetWriter的fileType决定
//        let pcmFilePath = outPutPath + "/outPut.wav"
        
        let outPutURL = URL(fileURLWithPath: pcmFilePath)
        
        do {
            assetWriter = try AVAssetWriter(outputURL: outPutURL, fileType: AVFileTypeWAVE)
        } catch {
            
            print(error)
            return nil
        }
        
        return assetWriter
    }
}
